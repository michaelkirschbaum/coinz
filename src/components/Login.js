/** @jsx jsx */
import React, { useState } from 'react'
import { Redirect, withRouter } from 'react-router-dom'
import PropTypes from 'prop-types'
import { css, jsx } from '@emotion/core'
import styled from '@emotion/styled'
// import logo from '../../public/logo.jpg'

const Error = styled.span`
  color: #eb516d
`

const Login = ({ setUser, location }) => {
  const [address, setAddress] = useState('')
  const [redirectToReferrer, setRedirectToReferrer] = useState(false)
  const [errors, setErrors] = useState({})

  const validate = (address) => {
    let inputErrors = {}

    if (!address) inputErrors.address = 'field must be provided'

    return inputErrors
  }

  const handleSubmit = async (event) => {
    event.preventDefault()

    // clear errors
    setErrors({})

    const inputErrors = validate(address)
    if (Object.entries(inputErrors).length) {
      setErrors(inputErrors)

      return
    }

    setUser(address)
    setRedirectToReferrer(true)
  }

  if (redirectToReferrer)
    return <Redirect to={location.state.from} />

  return (
    <div css={css`
      display: grid;
      text-align: center;
      position: absolute;
      margin: auto;
      top: 0;
      right: 0;
      bottom: 0;
      left: 0;
      height: 150px;
      width: 200px;
    `}>
      <img alt='logo' />
      <div>Welcome to Jobcoin! Sign in using your address.</div>
      <form onSubmit={handleSubmit}>
        <label>
          <div>Jobcoin Address</div>
          <input
            type='text'
            name='address'
            onChange={e => setAddress(e.target.value)}
          />
        </label>
        {errors.address &&
          <Error>{errors.address}</Error>
        }
        <input type='submit' value='Sign In' />
      </form>
    </div>
  )
}

Login.propTypes = {
  setUser: PropTypes.func.isRequired,
}

export default withRouter(Login)