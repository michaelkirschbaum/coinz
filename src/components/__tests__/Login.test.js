import React from 'react'
import { mount } from 'enzyme'
import Login from 'components/Login'

describe('Login', () => {
  let login
  beforeEach(() => {
    const context = { user: '' }
    const setUser = jest.fn().mockImplementation(user => {
      context.user = user
    })
    const location = { state: {} }

    login = mount(
      <Login.WrappedComponent setUser={setUser} location={location} />,
      { context }
    )

    global.fetch = jest.fn().mockImplementation(() => Promise.resolve({
      json: () => Promise.resolve({})
    }))
  })

  afterEach(() => {
    global.fetch.mockClear()
  })

  it('should match the snapshot', () => {
    expect(login).toMatchSnapshot()
  })

  it('should display error when address is invalid', () => {
    const form = login.find('form').first()
    form.find('input').first().simulate('change', {
      target: {
        value: ''
      }
    })
    form.simulate('submit')
    expect(form.text()).toContain('field must be provided')
  })

  it('should authenticate on click', () => {
    const form = login.find('form').first()
    form.find('input').first().simulate('change', {
      target: {
        value: 'Alice'
      }
    })
    form.simulate('submit')
    expect(login.context().user).toEqual('Alice')

    setUser.mockClear()
  })
})