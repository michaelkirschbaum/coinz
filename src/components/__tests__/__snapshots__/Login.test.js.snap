// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login should match the snapshot 1`] = `
<Login
  location={
    Object {
      "state": Object {},
    }
  }
  setUser={[MockFunction]}
>
  <EmotionCssPropInternal
    __EMOTION_LABEL_PLEASE_DO_NOT_USE__="Login"
    __EMOTION_TYPE_PLEASE_DO_NOT_USE__="div"
    css={
      Object {
        "map": undefined,
        "name": "fkr038",
        "next": undefined,
        "styles": "
      display: grid;
      text-align: center;
      position: absolute;
      margin: auto;
      top: 0;
      right: 0;
      bottom: 0;
      left: 0;
      height: 150px;
      width: 200px;
    ",
      }
    }
  >
    <div
      className="css-1b2d89h-Login"
    >
      <img
        alt="logo"
      />
      <div>
        Welcome to Jobcoin! Sign in using your address.
      </div>
      <form
        onSubmit={[Function]}
      >
        <label>
          <div>
            Jobcoin Address
          </div>
          <input
            name="address"
            onChange={[Function]}
            type="text"
          />
        </label>
        <input
          type="submit"
          value="Sign In"
        />
      </form>
    </div>
  </EmotionCssPropInternal>
</Login>
`;
